Fix 'JVM too old' as bsp

MaciejG604 opened the pull request:
Fixes #2412

Root cause:
- our Bsp command didn't take into consideration the options set with `using directives`, so the `using jvm 19` could never be taken into consideration when downloading jvm for bloop
- adding a different version of the jvm to the build made [Scala CLI add a different `--release` flag to the project](https://github.com/VirtusLab/scala-cli/blob/d8a4d59308b4f65368fac36d68b720e425170469/modules/build/src/main/scala/scala/build/Build.scala#L1028), the build client gets notified about this fact with [OnBuildTargetDidChange: notification](https://build-server-protocol.github.io/docs/specification/#onbuildtargetdidchange-notification), in case of metals the response for it is to restart the build server thus the constant recompilation. 

EDIT: The `-release` flag should not be set if the jvm of bloop is too low, thus it should not result in the buildTargetChange notification. However one of our functions (here )did not get passed a compiler argument thus it didn't check if the '-release' flag should be set. This resulted in the flag being set and reset which resulted in the buildTargetChange notification.

Since we already preprocess sources with `CrossSources.forInputs` there's no significant additional time complexity of the solution proposed.

MaciejG604 committed:
NIT: Update scala-cli-src script

MaciejG604 committed:
Add meaningful 'minimum JVM version' to the bloop config

Gedochao commented on modules/cli/src/main/scala/scala/cli/commands/shared/SharedCompilationServerOptions.scala:
```diff
@@ -276,7 +276,7 @@ final case class SharedCompilationServerOptions(
       javaOpts =
         (if (bloopDefaultJavaOpts) baseConfig.javaOpts
          else Nil) ++ bloopJavaOpt ++ bloopDefaultJvmOptions(logger).getOrElse(Nil),
-      minimumBloopJvm = javaV.getOrElse(8),
```
weren't we set on allowing to run Bloop with older Java versions? I have a vague recollection it should still run on 11 🤔 
I could be wrong

MaciejG604 replied:
Right now, we very restricitvely keep the JVM for bloop at least version 17, so I thought using 17 as minimum here is reasonable.
If the minimum is not set there's a possibility that an existing bloop server won't be restarted if it's jvm is older than requested. Actually that was part of the problem of the issue.

Gedochao replied:
Alright then, let's bump to 17 👍

Gedochao commented:
Could we add an anti-regression test for this?
should be possible to set up an integration test with BSP, I think

MaciejG604 committed:
Extract initial BspReloadableOptions from sources in Bsp command

MaciejG604 committed:
Restart bloop on workspace/reload when its JVM version is too low

Add test for restarting bloop with the correct JVM version after workspace/reload

Add test for starting bsp while bloop is running with JVM version too low

Change the warning message

Gedochao commented on scala-cli-src:
```diff
@@ -1,5 +1,5 @@
 set -euo pipefail
 SCRIPT_DIR="$( cd "$( dirname "$0" )" && pwd )"
-LAUNCHER="$(cd "$SCRIPT_DIR" && ./mill show cli.launcher </dev/null | jq -r . | sed 's/ref:[a-z0-9]*://')"
+LAUNCHER="$(cd "$SCRIPT_DIR" && ./mill show cli.launcher </dev/null | jq -r . | sed 's/ref:[a-z0-9]*:[a-z0-9]*://')"
```
?

Gedochao replied:
right... `scala-cli-src` was broken somewhere along the way, since `./mill show cli.launcher` output changed.
ok.
